wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,229 @@
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiFunction;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.rds.RdsClient;
+import software.amazon.awssdk.services.rds.model.DBClusterEndpoint;
+import software.amazon.awssdk.services.rds.model.DescribeDbClusterEndpointsResponse;
+import software.amazon.awssdk.services.rds.model.Filter;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.util.CacheMap;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+
+public class CustomEndpointMonitorImpl implements CustomEndpointMonitor {
+  private static final Logger LOGGER = Logger.getLogger(CustomEndpointPlugin.class.getName());
+  // Keys are custom cluster endpoint identifiers, values are information objects for the associated custom cluster.
+  protected static final CacheMap<String, CustomEndpointInfo> customEndpointInfoCache = new CacheMap<>();
+
+  protected final AtomicBoolean stop = new AtomicBoolean(false);
+  protected final RdsUtils rdsUtils = new RdsUtils();
+  protected final RdsClient rdsClient;
+  protected final HostSpec customEndpointHostSpec;
+  protected final String endpointIdentifier;
+  protected final Region region;
+  protected final long refreshRateNano;
+  protected final long cacheEntryExpirationNano;
+
+  protected final List<PluginService> pluginServices = new LinkedList<>();
+  protected final ReadWriteLock pluginServicesLock = new ReentrantReadWriteLock();
+  // TODO: is it problematic for each monitor thread to have its own executor service
+  protected final ExecutorService monitorExecutor = Executors.newSingleThreadExecutor(runnableTarget -> {
+    final Thread monitoringThread = new Thread(runnableTarget);
+    monitoringThread.setDaemon(true);
+    if (!StringUtils.isNullOrEmpty(monitoringThread.getName())) {
+      monitoringThread.setName(monitoringThread.getName() + "-CustomEndpointMonitor");
+    }
+    return monitoringThread;
+  });
+
+  public CustomEndpointMonitorImpl(
+      HostSpec customClusterHostSpec,
+      Properties props,
+      BiFunction<HostSpec, Region, RdsClient> rdsClientFunc,
+      long cacheEntryExpirationNano) {
+    this.customEndpointHostSpec = customClusterHostSpec;
+    this.region = getRegion(customClusterHostSpec, props);
+    this.rdsClient = rdsClientFunc.apply(customClusterHostSpec, this.region);
+    this.cacheEntryExpirationNano = cacheEntryExpirationNano;
+
+    this.refreshRateNano = TimeUnit.MILLISECONDS.toNanos(CUSTOM_ENDPOINT_INFO_REFRESH_RATE.getLong(props));
+
+    this.endpointIdentifier = this.rdsUtils.getRdsClusterId(customClusterHostSpec.getHost());
+    if (StringUtils.isNullOrEmpty(this.endpointIdentifier)) {
+      throw new RuntimeException(
```

Comment:
I'm not sure RuntimeException. Should we consider SQLException?

Reply:
Sure. I noticed we are throwing RuntimeExceptions instead of SQLExceptions in some other areas of the code too, for example in the EFM plugin when we have trouble getting the node ID. Its a bit unclear to me how we're deciding when to use RuntimeExceptions instead of SQLExceptions - any tips?

Reply:
The reason to checked `SQLException` is that a user app catch them and can handle them. Unchecked exceptions are mostly for some unpredictable/unhandled cases.

In this particular case I'd suggest to wrap logic inside `createMonitorIfAbsent` in try-catch. You can catch `RutimeException` and convert it to `SQLException` and raise.

Reply:
I moved the parsing logic outside the computeIfAbsent lambda now, so it is now a SQLException instead of RuntimeException